File: MapDemo/Objective C/CovidLayersExample.m
//
//  CovidLayersExample.m
//  MapDemo
//
//  Created by apple on 04/06/20.
//  Copyright © 2020 MMI. All rights reserved.
//

#import "CovidLayersExample.h"
#import "MMIInteractiveLayersTableViewController.h"

@interface CovidLayersExample () <MapmyIndiaMapViewDelegate, MMIInteractiveLayersTableViewControllerDelegate>

@property (nonatomic) IBOutlet MapmyIndiaMapView *mapView;

@property (weak, nonatomic) IBOutlet UIButton *covidMarkerToggleButton;

@property (weak, nonatomic) IBOutlet UIButton *covid19Button;

@property (weak, nonatomic) IBOutlet UILabel *covidInfoLabel;


@end

@implementation CovidLayersExample

- (IBAction)covidMarkerToggleButtonPressed:(UIButton *)sender {
    BOOL newState = !_covidMarkerToggleButton.isSelected;
    [_covidMarkerToggleButton setSelected:newState];
    [_mapView setShouldShowPopupForInteractiveLayer:newState];
}

- (IBAction)covid19ButtonPressed:(UIButton *)sender {
    MMIInteractiveLayersTableViewController * interactiveLayersController = [[MMIInteractiveLayersTableViewController alloc] initWithNibName:nil bundle:nil];
    interactiveLayersController.interactiveLayers = self.mapView.interactiveLayers;
    interactiveLayersController.selectedInteractiveLayers = self.mapView.visibleInteractiveLayers;
    interactiveLayersController.delegate = self;
    UINavigationController *wrapper = [[UINavigationController alloc] initWithRootViewController:interactiveLayersController];
    wrapper.navigationBar.tintColor = self.navigationController.navigationBar.tintColor;
    [self.navigationController presentViewController:wrapper animated:YES completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _mapView.delegate = self;
    
    _covidInfoLabel.text = @"";
    _covidInfoLabel.backgroundColor = [UIColor whiteColor];
    _covidInfoLabel.textAlignment = NSTextAlignmentCenter;
    
    [_covidMarkerToggleButton setSelected:self.mapView.shouldShowPopupForInteractiveLayer];
    [_covid19Button setHidden:YES];
}


- (void)mapView:(MapmyIndiaMapView *)mapView authorizationCompleted:(BOOL)isSuccess
{
    if(isSuccess) {
        [self.mapView getCovidLayers];
    }
}

- (void)mapViewInteractiveLayersReady:(MapmyIndiaMapView *)mapView
{
    if (self.mapView.interactiveLayers && self.mapView.interactiveLayers.count > 0) {
        [_covid19Button setHidden:NO];
    }
}

- (void)layersSelected:(NSArray<MapmyIndiaInteractiveLayer *> *)selectedInteractiveLayers
{
    for (MapmyIndiaInteractiveLayer *layer in self.mapView.interactiveLayers) {
        BOOL isSelected = NO;
        for (MapmyIndiaInteractiveLayer *selectedLayer in selectedInteractiveLayers) {
            if (selectedLayer.layerId == layer.layerId) {
                isSelected = YES;
            }
        }
        
        if (isSelected) {
            [self.mapView showInteractiveLayerOnMapForLayerId:layer.layerId];
        } else {
            [self.mapView hideInteractiveLayerFromMapForLayerId:layer.layerId];
        }
    }
}

- (void)didDetectCovidInfo:(MapmyIndiaCovidInfo *)covidInfo
{
    if (covidInfo) {
        NSMutableArray<NSString *> * covidInfoText = [[NSMutableArray alloc] init];
        
        if (covidInfo.total) {
            [covidInfoText addObject: [NSString stringWithFormat:@"Total: %@", covidInfo.total]];
        }
        if (covidInfo.cured) {
            [covidInfoText addObject: [NSString stringWithFormat:@"Cured: %@", covidInfo.cured]];
        }
        if (covidInfo.death) {
            [covidInfoText addObject: [NSString stringWithFormat:@"Death: %@", covidInfo.death]];
        }
        if (covidInfo.confInd) {
            [covidInfoText addObject: [NSString stringWithFormat:@"ConfInd: %@", covidInfo.confInd]];
        }
        if (covidInfo.areaZone) {
            [covidInfoText addObject: [NSString stringWithFormat:@"Zone: %@", covidInfo.areaZone]];
        }
        if (covidInfo.districtName) {
            [covidInfoText addObject: [NSString stringWithFormat:@"District: %@", covidInfo.districtName]];
        }
        if (covidInfo.stateName) {
            [covidInfoText addObject: [NSString stringWithFormat:@"State: %@", covidInfo.stateName]];
        }
        
        _covidInfoLabel.text =  [covidInfoText componentsJoinedByString:@"\n"];
    } else {
        _covidInfoLabel.text =  @"";
    }
}

@end

File: MapDemo/Swift/CovidLayersExample.swift
//
//  CovidLayersExample.swift
//  MapDemo
//
//  Created by apple on 04/06/20.
//  Copyright © 2020 MMI. All rights reserved.
//

import UIKit
import MapmyIndiaMaps

@objc(CovidLayersExample_Swift)
class CovidLayersExample_Swift: UIViewController {

    @IBOutlet weak var mapView: MapmyIndiaMapView!
    @IBOutlet weak var covidInfoLabel: UILabel!
    
    @IBOutlet weak var covidMarkerToggleButton: UIButton!
    @IBOutlet weak var covid19Button: UIButton!
    
    
    @IBAction func covid19ButtonPressed(_ sender: UIButton) {
        let vc = CovidLayersTableVC(nibName: nil, bundle: nil)
        vc.interactiveLayers = self.mapView.interactiveLayers ?? [MapmyIndiaInteractiveLayer]()
        vc.selectedInteractiveLayers = self.mapView.visibleInteractiveLayers ?? [MapmyIndiaInteractiveLayer]()
        vc.delegate = self
        let navVC = UINavigationController(rootViewController: vc)
        present(navVC, animated: true, completion: nil)
    }
    
    @IBAction func covidMarkerToggleButtonPressed(_ sender: UIButton) {
        let newState = !sender.isSelected;
        covidMarkerToggleButton.isSelected = newState
        self.mapView.shouldShowPopupForInteractiveLayer = newState
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        covidInfoLabel.text = ""
        covidInfoLabel.backgroundColor = .white
        covidInfoLabel.textAlignment = .center
        
        mapView.delegate = self
        covid19Button.isHidden = true
    }
}

extension CovidLayersExample_Swift: CovidLayersTableVCDelegate {
    func layersSelected(selectedInteractiveLayers: [MapmyIndiaInteractiveLayer]) {
        if let allLayers = self.mapView.interactiveLayers {
            for layer in allLayers {
                var isSelected = false
                if let layerId = layer.layerId {
                    for selectedLayer in selectedInteractiveLayers {
                        if let selectedLayerId = selectedLayer.layerId {
                            if selectedLayerId == layerId {
                                isSelected = true
                            }
                        }
                    }
                                
                    if isSelected {
                        self.mapView.showInteractiveLayerOnMap(forLayerId: layerId)
                    } else {
                        self.mapView.hideInteractiveLayerFromMap(forLayerId: layerId)
                    }
                }
            }
        }
    }
}

extension CovidLayersExample_Swift: MapmyIndiaMapViewDelegate {
    func mapView(_ mapView: MapmyIndiaMapView, authorizationCompleted isSuccess: Bool) {
        if isSuccess {
            self.mapView.getCovidLayers()
        }
    }
    
    func mapViewInteractiveLayersReady(_ mapView: MapmyIndiaMapView) {
        if self.mapView.interactiveLayers?.count ?? 0 > 0 {
            covid19Button.isHidden = false
        }
    }
    
    func didDetect(_ covidInfo: MapmyIndiaCovidInfo?) {
        if let covidInfo = covidInfo {
            var covidInfoText = [String]()
            
            if let total = covidInfo.total {
                covidInfoText.append("Total: \(total)")
            }
            if let cured = covidInfo.cured {
                covidInfoText.append("Cured: \(cured)")
            }
            if let death = covidInfo.death {
                covidInfoText.append("Death: \(death)")
            }
            if let confInd = covidInfo.confInd {
                covidInfoText.append("ConfInd: \(confInd)")
            }
            if let areaZone = covidInfo.areaZone {
                covidInfoText.append("Zone: \(areaZone)")
            }
            if let districtName = covidInfo.districtName {
                covidInfoText.append("District: \(districtName)")
            }
            if let stateName = covidInfo.stateName {
                covidInfoText.append("State: \(stateName)")
            }
            
            
            covidInfoLabel.text = covidInfoText.joined(separator: "\n")
        } else {
            covidInfoLabel.text = ""
        }
    }
}
